src/compiletest.md

```diff
@@ -0,0 +1,183 @@
+optional.
+
+Parsers will override a given header command property's default value merely by being specified in the test file as a header
+command or by having a parameter value specified in the test file, depending on the header command.
+
+Parsers defined in `impl Config` are typically named `parse_<header_command>` (note kebab-case `<header-command>` transformed
+to snake-case `<header_command>`).  `impl Config` also defines several 'low-level' parsers which make it simple to parse
+common patterns like simple presence or not (`parse_name_directive()`), header-command:parameter(s)
+(`parse_name_value_directive()`), optional parsing only if a particular `cfg` attribute is defined (`has_cfg_prefix()`) and
+many more.  The low-level parsers are found near the end of the `impl Config` block--be sure to look through them and their
+associated parsers immediately above to see how they are used to avoid writing additional parsing code unneccessarily.
+
+As a concrete example, here is the implementation for the `parse_failure_status()` parser, in
+`src/tools/compiletest/src/header.rs`:
+```diff
+@@ -232,6 +232,7 @@ pub struct TestProps {
+     // customized normalization rules
+     pub normalize_stdout: Vec<(String, String)>,
+     pub normalize_stderr: Vec<(String, String)>,
++    pub failure_status: i32,
+ }
+ 
+ impl TestProps {
+@@ -260,6 +261,7 @@ impl TestProps {
+             run_pass: false,
+             normalize_stdout: vec![],
+             normalize_stderr: vec![],
++            failure_status: 101,
+         }
+     }
+ 
+@@ -383,6 +385,10 @@ impl TestProps {
+             if let Some(rule) = config.parse_custom_normalization(ln, "normalize-stderr") {
+                 self.normalize_stderr.push(rule);
+             }
++
++            if let Some(code) = config.parse_failure_status(ln) {
++                self.failure_status = code;
++            }
+         });
+ 
+         for key in &["RUST_TEST_NOCAPTURE", "RUST_TEST_THREADS"] {
+@@ -488,6 +494,13 @@ impl Config {
+         self.parse_name_directive(line, "pretty-compare-only")
+     }
+ 
++    fn parse_failure_status(&self, line: &str) -> Option<i32> {
++        match self.parse_name_value_directive(line, "failure-status") {
++            Some(code) => code.trim().parse::<i32>().ok(),
++            _ => None,
++        }
++    }
+```
+
+## Implementing the behavior change
+When a test invokes a particular header command, it is expected that some behavior will change as a result.  What behavior,
+obviously, will depend on the purpose of the header command.  In the case of `failure-status`, the behavior that changes
+is that `compiletest` expects the failure code defined by the header command invoked in the test, rather than the default
+value.
+
+Although specific to `failure-status` (as every header ommand will have a different implementation in order to invoke
+behavior change) perhaps it is helpful to see the behavior hange implementation of one case, simply as an example.  To
```

Comment:
"hange" -> "change"

Reply:
weird; why am I leaving out so many first letters?  :)

Reply:
And they are both "c"...